File: 2244-minimum-rounds-to-complete-all-tasks/2244-minimum-rounds-to-complete-all-tasks.cpp
#include <iostream>
#include <vector>
#include <map>
class Solution
{
public:
    int minimumRounds(std::vector<int> &tasks)
    {
        std::map<int, uint32_t> hm;
        for (size_t i = 0; i < tasks.size(); i++)
        {
            auto count = hm.find(tasks[i]);
            if (count == hm.end())
                hm.insert({tasks[i], 1});
            else
                count->second += 1;
        }
        int ans = 0;
        for (auto const &[_, v] : hm)
        {
            if (v == 1)
                return -1;
            auto rem = v % 3;
            ans += v / 3;
            if (rem != 0)
                ans += 1;
        }
        return ans;
    }
};

File: 2244-minimum-rounds-to-complete-all-tasks/2244-minimum-rounds-to-complete-all-tasks.rs
use std::collections::HashMap;

impl Solution {
    pub fn minimum_rounds(tasks: Vec<i32>) -> i32 {
        let mut hm: HashMap<i32, u32> = Default::default();
        for x in tasks {
            *hm.entry(x).or_default() += 1;
        }
        let mut res = 0;
        for (_, c) in hm {
            match Self::div_rem(c, 3) {
                (t, 0) => {res += t;}
                (0, 1) => {return -1;}
                (t, _) => {res += t + 1;}
            }
        }
        res as i32
    }

    fn div_rem<T>(x: T, y: T) -> (T, T) where T: std::ops::Div<Output=T> + std::ops::Rem<Output=T> + Copy {
        let quot = x / y;
        let rem = x % y;
        (quot, rem)
    }
}